src/config_diff.ml

```diff
@@ -229,3 +245,110 @@ let trim_tree left right =
     let trees = make_diff_trees left right in
     diff [] (trim_trees trees) [(Option.some left, Option.some right)];
     !(trees.sub)
+
+(* the following builds a diff_func to return a unified diff string of
+   configs or config commands
+ *)
+let list_but_last l =
+    let len = List.length l in
+    List.filteri (fun i _ -> i < len - 1) l
+
+let ppath_to_string_if_new (c: string list ref) (path: string list) =
+    let p = list_but_last path in
+    if (!c <> p) then
+         (c := p; "[" ^ (String.concat " " !c) ^ "]" ^ "\n") else ""
```

Comment:
Is there a significant performance concern with `Printf.sprintf "[%s]\n" (String.concat " " !c)`? I think it would be more readable.

Reply:
Quite right, regarding the second point; I'll fix it. Regarding the first point, I plan to drop all mutable references soon, for that very reason --- this will be done at the same time: this example, and the underlying use of mutable trees for collecting the difference trees. All state will be passed in return values, as is properly functional.